static/js/tools/map/chart_loader.tsx

```diff
@@ -133,6 +138,140 @@ export function ChartLoader(): JSX.Element {
   //   };
   // }, []);
 
+  // +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+  // START useReducer
+  const [chartStore, dispatch] = useReducer(chartStoreReducer, emptyChartStore);
+  // END useReducer
+  // -------------------------------------------------------------------------
+
+  // +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+  // The functions below are memoized and only change when the dependant
+  // variables change.
+  const geoJsonReady = useCallback(() => {
+    const c = chartStore.geoJson.context;
+    return (
+      !_.isEmpty(c) &&
+      placeInfo.value.enclosingPlace.dcid == c.placeInfo.enclosingPlace.dcid &&
+      placeInfo.value.enclosedPlaceType == c.placeInfo.enclosedPlaceType
+    );
+  }, [
+    chartStore.geoJson.context,
+    placeInfo.value.enclosingPlace.dcid,
+    placeInfo.value.enclosedPlaceType,
+  ]);
+
+  const defaultStatReady = useCallback(() => {
+    const c = chartStore.defaultStat.context;
+    return (
+      !_.isEmpty(c) &&
+      placeInfo.value.enclosingPlace.dcid == c.placeInfo.enclosingPlace.dcid &&
+      placeInfo.value.enclosedPlaceType == c.placeInfo.enclosedPlaceType &&
+      statVar.value.dcid == c.statVar.dcid &&
+      statVar.value.date == c.statVar.date
+    );
+  }, [
+    chartStore.defaultStat.context,
+    placeInfo.value.enclosingPlace.dcid,
+    placeInfo.value.enclosedPlaceType,
+    statVar.value.dcid,
+    statVar.value.date,
+  ]);
+
+  const allStatReady = useCallback(() => {
+    const c = chartStore.allStat.context;
+    return (
+      !_.isEmpty(c) &&
+      placeInfo.value.enclosingPlace.dcid == c.placeInfo.enclosingPlace.dcid &&
+      placeInfo.value.enclosedPlaceType == c.placeInfo.enclosedPlaceType &&
+      statVar.value.dcid == c.statVar.dcid &&
+      statVar.value.date == c.statVar.date
+    );
+  }, [
+    chartStore.allStat.context,
+    placeInfo.value.enclosingPlace.dcid,
+    placeInfo.value.enclosedPlaceType,
+    statVar.value.dcid,
+    statVar.value.date,
+  ]);
+
+  const denomStatReady = useCallback(() => {
+    const c = chartStore.denomStat.context;
+    console.log("denomStatReady");
+    return (
+      !_.isEmpty(c) &&
+      placeInfo.value.enclosingPlace.dcid == c.placeInfo.enclosingPlace.dcid &&
+      placeInfo.value.enclosedPlaceType == c.placeInfo.enclosedPlaceType &&
+      statVar.value.denom == c.statVar.denom
+    );
+  }, [
+    chartStore.denomStat.context,
+    placeInfo.value.enclosingPlace.dcid,
+    placeInfo.value.enclosedPlaceType,
+    statVar.value.denom,
+  ]);
+
+  const breadcrumbStatReady = useCallback(() => {
+    const c = chartStore.breadcrumbStat.context;
+    console.log("breadcrumbStatReady");
```

Comment:
are these console.logs needed? because I notice some of these callbacks have console logs and some don't

Reply:
removed them now